### PR TITLE
Fix uart async test

### PR DESCRIPTION
**Pull request by shlomow**
This pr fixes multiple bugs in the uart async test.

### CODE REVIEW

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 1]
```diff
@@ -74,13 +74,13 @@ void test_single_read(void)
 			  "Initial buffer check failed");
 
 	uart_rx_enable(uart_dev, rx_buf, 10, 50);
-	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(100)), -EAGAIN,
+	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(10)), -EAGAIN,
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:77]: At least on nrf drivers RX timeout was implemented as break between bytes. It allows to report user when some data was received. It seems like more reasonable than timeouting even without any bytes received. In that case 100ms is an arbitrary value.

However, i see that api is not clear about it.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 2]
```diff
@@ -74,13 +74,13 @@ void test_single_read(void)
...
 		      "RX_RDY not expected at this point");
 
 	uart_tx(uart_dev, tx_buf, sizeof(tx_buf), 100);
 	zassert_equal(k_sem_take(&tx_done, K_MSEC(100)), 0, "TX_DONE timeout");
 	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(100)), 0, "RX_RDY timeout");
-	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(100)), -EAGAIN,
+	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(10)), -EAGAIN,
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:83]: 100ms was an arbitrary value.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 3]
```diff
@@ -423,8 +423,7 @@ void test_forever_timeout(void)
 	zassert_not_equal(k_sem_take(&tx_aborted, K_MSEC(1000)), 0,
 			  "TX_ABORTED timeout");
 	zassert_equal(k_sem_take(&tx_done, K_MSEC(100)), 0, "TX_DONE timeout");
-	zassert_not_equal(k_sem_take(&rx_rdy, K_MSEC(1000)), 0,
-			  "RX_RDY timeout");
+	zassert_equal(k_sem_take(&rx_rdy, K_MSEC(100)), 0, "RX_RDY timeout");
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:426]: when rx is started with `SYS_FOREVER_MS` timeout, it shall report RX_RDY event only if full buffer is recieved (100 bytes) so there should be no event after 5 bytes.

File: tests/drivers/uart/uart_async_api/src/test_uart_async.c [hunk 4]
```diff
@@ -564,7 +564,7 @@ void test_long_buffers(void)
 	memset(long_rx_buf, 0, sizeof(long_rx_buf));
 	memset(long_tx_buf, 1, sizeof(long_tx_buf));
 
-	uart_rx_enable(uart_dev, long_rx_buf, sizeof(long_rx_buf), 10);
+	uart_rx_enable(uart_dev, long_rx_buf, sizeof(long_rx_buf), 150);
```
Comment (nordic-krch) [inline tests/drivers/uart/uart_async_api/src/test_uart_async.c:567]: RX timeout is a inactive time on RX which leads to RX_RDY event.

**Comment by nordic-krch** [comment]
@shlomow i think that test failures you are seeing are caused by different interpretation of rx timeout. I will make a PR to clarify what rx timeout means.

**Comment by nordic-krch** [comment]
See https://github.com/zephyrproject-rtos/zephyr/blob/cc470a2bd6b961fa10a865a996560da3d9f50be4/include/drivers/uart.h#L56 where it is clarified.

**Comment by shlomow** [comment]
@nordic-krch Thanks a lot for the clarifications! I think that I expected the driver to notify the user for available data as soon as it detects it. Since I am using stm32 the hardware has a feature that it sends interrupt when it finds no activity for one uart frame (typically 10 bit times). This api force some latency on users who want to respond fast to uart reception. I guess this is what I missed. Thanks again!

**Comment by nordic-krch** [comment]
@shlomow some time ago i was also mentioning that milliseconds timeout is too much in some cases when API was initially discussed: https://github.com/zephyrproject-rtos/zephyr/pull/10820#discussion_r230386184
As there are only few implementations present maybe it's worth to pursue change to microseconds.

**Comment by nordic-krch** [comment]
@shlomow I've created #30861. Can you leave your comment there?

**Comment by shlomow** [comment]
I commented there. Thanks again! I am closing this PR.

**Pull request closed without merging**